pkg/services/sqlstore/datasource.go

```diff
@@ -121,9 +131,13 @@ func AddDataSource(cmd *models.AddDataSourceCommand) error {
 			Updated:           time.Now(),
 			Version:           1,
 			ReadOnly:          cmd.ReadOnly,
+			Uid:               cmd.Uid,
 		}
 
 		if _, err := sess.Insert(ds); err != nil {
+			if dialect.IsUniqueConstraintViolation(err) && strings.Contains(strings.ToLower(dialect.ErrorMessage(err)), "uid") {
```

Comment:
I agree with @bergquist.

Reply:
True but problem is that the DB errors for Mysql and sqlite do not have any more info. So could be moved to dialect and made more robust for PG but for others this would basically be the same code.